transaction/std/src/transaction_builder.rs

```diff
@@ -280,12 +284,205 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
+                    .try_into()
+                    .expect("size was checked earlier");
+                let new_masked_amount = MaskedAmountV2::new_from_amount_shared_secret(
+                    real_amount,
+                    amount_shared_secret,
+                )?;
+                let (new_amount, blinding) =
+                    new_masked_amount.get_value_from_amount_shared_secret(amount_shared_secret)?;
+                debug_assert!(real_amount == new_amount);
+
+                real_tx_out.masked_amount = Some(MaskedAmount::V2(new_masked_amount));
+
+                let output_secret = OutputSecret {
+                    amount: new_amount,
+                    blinding,
+                };
+
+                self.outputs_and_secrets.push((real_tx_out, output_secret));
+
+                Ok(real_amount)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // Add the real fractional change output
+        {
+            let mut real_change = fractional_change.tx_out.clone();
+
+            let amount_shared_secret: &[u8; 32] = &fractional_change.amount_shared_secret[..]
+                .try_into()
+                .expect("size was checked earlier");
+            let new_masked_amount = MaskedAmountV2::new_from_amount_shared_secret(
+                sci_change_amount,
+                amount_shared_secret,
+            )?;
+            let (new_amount, blinding) =
+                new_masked_amount.get_value_from_amount_shared_secret(amount_shared_secret)?;
+            debug_assert!(sci_change_amount == new_amount);
+
+            real_change.masked_amount = Some(MaskedAmount::V2(new_masked_amount));
+
+            let output_secret = OutputSecret {
+                amount: new_amount,
+                blinding,
+            };
+
+            self.outputs_and_secrets.push((real_change, output_secret));
+        }
+
+        // Enforce all non-partial fill rules so that our transaction will be valid
+        if rules.required_outputs.len() != sci.required_output_amounts.len() {
+            return Err(SignedContingentInputError::WrongNumberOfRequiredOutputAmounts);
+        }
+        // 1. Required outputs
+        for (required_output, unmasked_amount) in rules
+            .required_outputs
+            .iter()
+            .zip(sci.required_output_amounts.iter())
+        {
+            // Check if the required output is already there
+            if !self
+                .outputs_and_secrets
+                .iter()
+                .any(|(output, _sec)| output == required_output)
+            {
+                // If not, add it
+                self.outputs_and_secrets
+                    .push((required_output.clone(), unmasked_amount.clone().into()));
+            }
+        }
+        // 2. Max tombstone block
+        if rules.max_tombstone_block != 0 {
+            self.impose_tombstone_block_limit(rules.max_tombstone_block);
+        }
```

Comment:
:wrench: This looks suspiciously like a copy and paste from `add_presigned_input()`.  Is there a reason this didn't get extracted out to a common method?  It seems likely that as more rule types get added these become out of sync.  Even now there doesn't appear to be tests to ensure this code does what it's supposed to.

Reply:
I think you are right -- we should extract a common method here.

I also think since I envision that people may use "combined" SCIs that have required and fractional outputs, I should probably have a test case like that.